Framework/CurveFitting/src/Algorithms/DoublePulseFit.cpp

```diff
@@ -0,0 +1,406 @@
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidCurveFitting/Algorithms/DoublePulseFit.h"
+#include "MantidCurveFitting/CostFunctions/CostFuncFitting.h"
+
+#include "MantidAPI/CompositeFunction.h"
+#include "MantidAPI/FuncMinimizerFactory.h"
+#include "MantidAPI/FunctionFactory.h"
+#include "MantidAPI/IFuncMinimizer.h"
+#include "MantidAPI/ITableWorkspace.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/TableRow.h"
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidAPI/WorkspaceGroup.h"
+#include "MantidCurveFitting/Functions/DeltaFunction.h"
+
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/Exception.h"
+#include "MantidKernel/StartsWithValidator.h"
+
+#include <math.h>
+#include <memory>
+
+using namespace Mantid::CurveFitting::Functions;
+
+namespace Mantid {
+namespace CurveFitting {
+namespace Algorithms {
+
+void setMultiDataProperties(const Mantid::API::IAlgorithm &qensFit,
+                            Mantid::API::IAlgorithm &fit,
+                            const Mantid::API::MatrixWorkspace_sptr &workspace,
+                            const std::string &suffix) {
+  fit.setProperty("InputWorkspace" + suffix, workspace);
+
+  int workspaceIndex = qensFit.getProperty("WorkspaceIndex" + suffix);
+  fit.setProperty("WorkspaceIndex" + suffix, workspaceIndex);
+
+  double startX = qensFit.getProperty("StartX" + suffix);
+  double endX = qensFit.getProperty("EndX" + suffix);
+  fit.setProperty("StartX" + suffix, startX);
+  fit.setProperty("EndX" + suffix, endX);
+
+  std::vector<double> exclude = qensFit.getProperty("Exclude" + suffix);
+  fit.setProperty("Exclude" + suffix, exclude);
+}
+
+void setMultiDataProperties(
+    const Mantid::API::IAlgorithm &qensFit, Mantid::API::IAlgorithm &fit,
+    const std::vector<Mantid::API::MatrixWorkspace_sptr> &workspaces) {
+  setMultiDataProperties(qensFit, fit, workspaces[0], "");
+
+  for (auto i = 1u; i < workspaces.size(); ++i)
+    setMultiDataProperties(qensFit, fit, workspaces[i],
+                           "_" + std::to_string(i));
+}
+
+Mantid::API::IFunction_sptr
+getDoublePulseFunction(std::shared_ptr<const API::ParamFunction> const function,
+                       double offset, double firstPulseWeight,
+                       double secondPulseWeight) {
+  auto clonedFunction = function->clone();
+
+  auto convolution =
+      std::make_shared<Mantid::CurveFitting::Functions::Convolution>();
+  auto delta1 =
+      std::make_shared<Mantid::CurveFitting::Functions::DeltaFunction>();
+  auto delta2 =
+      std::make_shared<Mantid::CurveFitting::Functions::DeltaFunction>();
+  auto deltaComposite = std::make_shared<Mantid::API::CompositeFunction>();
+
+  convolution->setAttributeValue("FixResolution", false);
+  delta1->setParameter("Centre", -offset);
```

Comment:
another way we can do this is:
```suggestion
  std::vector<double> centres = {-offset, 0.0};
  std::vector<double> height = {firstPulseWeight, secondPulseWeight};
  for(j=0;j<height.size();j++){
      delta = std::make_shared<Mantid::CurveFitting::Functions::DeltaFunction>();
      delta->setParameter("Centre", centre[j]);
      delta->setParameter("Height", height[j]);
      delta->fixAll();
      deltaComposite->addFunction(delta);
  }
```

Reply:
I don't think you gain much in simplicity adding a loop when there are only two items. But acknowledge this could go either way.